ingest/ingest-projects.py

```diff
@@ -0,0 +1,374 @@
+    def orElse(self, action):
+        return Maybe.of(action()) if self.value is None else Maybe.of(self.value)
+
+    def do(self, action):
+        if self.value:
+            action(self.value)
+        return self
+
+    def filter(self, action):
+        return Maybe.of(filter(action, self.value)) if self.value is not None else Maybe.nothing()
+
+    def followedBy(self, action):
+        return self.andThen(lambda _: action)
+
+    def one(self):
+        try:
+            return Maybe.of(next(self.value)) if self.value is not None else Maybe.nothing()
+        except StopIteration:
+            return Maybe.nothing()
+        except TypeError:
+            return self
+
+    def list(self):
+        return list(self.value) if self.value is not None else []
+
+
+def load_file(filepath):
+    with open(filepath) as _file:
+        return _file.read().replace('\n', " ")
+
+
+get_projects_query = load_file("queries/listProjects.rq")
+describe_project_query = load_file("queries/describeProject.rq")
+
+PROV = Namespace("http://www.w3.org/ns/prov#")
+BIBO = Namespace("http://purl.org/ontology/bibo/")
+VCARD = Namespace("http://www.w3.org/2006/vcard/ns#")
+VIVO = Namespace('http://vivoweb.org/ontology/core#')
+VITRO = Namespace("http://vitro.mannlib.cornell.edu/ns/vitro/0.7#")
+VITRO_PUB = Namespace("http://vitro.mannlib.cornell.edu/ns/vitro/public#")
+OBO = Namespace("http://purl.obolibrary.org/obo/")
+DCO = Namespace("http://info.deepcarbon.net/schema#")
+FOAF = Namespace("http://xmlns.com/foaf/0.1/")
+
+
+def get_metadata(id):
+    return {"index": {"_index": "dco", "_type": "project", "_id": id}}
+
+
+def get_id(dco_id):
+    return dco_id[dco_id.rfind('/') + 1:]
+
+
+def select(endpoint, query):
+    sparql = SPARQLWrapper(endpoint)
+    sparql.setQuery(query)
+    sparql.setReturnFormat(JSON)
+    results = sparql.query().convert()
+    return results["results"]["bindings"]
+
+
+def describe(endpoint, query):
+    sparql = SPARQLWrapper(endpoint)
+    sparql.setQuery(query)
+    try:
+        return sparql.query().convert()
+    except RuntimeWarning:
+        pass
+
+
+def get_projects(endpoint):
+    r = select(endpoint, get_projects_query)
+    return [rs["project"]["value"] for rs in r]
+
+
+def process_project(project, endpoint):
+    prj = create_project_doc(project=project, endpoint=endpoint)
+    if "dcoId" in prj and prj["dcoId"] is not None:
+        return [json.dumps(get_metadata(get_id(prj["dcoId"]))), json.dumps(prj)]
+    else:
+        return []
+
+
+def describe_project(endpoint, project):
+    q = describe_project_query.replace("?project", "<" + project + ">")
+    return describe(endpoint, q)
+
+def get_thumbnail(project):
+    return Maybe.of(project).stream() \
+        .flatmap(lambda p: p.objects(VITRO_PUB.mainImage)) \
+        .flatmap(lambda i: i.objects(VITRO_PUB.thumbnailImage)) \
+        .flatmap(lambda t: t.objects(VITRO_PUB.downloadLocation)) \
+        .map(lambda t: t.identifier) \
+        .one().value
+
```

Comment:
In the comment for this method please add a reference to the mapping file so that we can see what it is we're creating.